### ISSUE TITLE
UnicodeDecodeError exception in version 2.4.5 (windows specific)

### ISSUE THREAD

**Issue by anton-tagunov** [buildout/buildout#279]
This exception is raised during reading file develop\tmp0nku9vrtbuild\ **backend.exe**
```
> bin\buildout
Upgraded:
  zc.buildout version 2.4.5;
restarting.
Generated script 'xxxx\\bin\\buildout'.
Develop: 'xxxxx\\backend'
While:
  Installing.
  Processing develop directory 'D:\\project\\boss\\boss\\backend'.

An internal error occurred due to a bug in either zc.buildout or in a
recipe being used:
Traceback (most recent call last):
  File "xxxx\.buildout\eggs\zc.buildout-2.4.5-py3.4.egg\zc\buildout\buildout.py", line 1992, in main
    getattr(buildout, command)(args)
  File "xxxx\.buildout\eggs\zc.buildout-2.4.5-py3.4.egg\zc\buildout\buildout.py", line 528, in install
    installed_develop_eggs = self._develop()
  File "xxxx\.buildout\eggs\zc.buildout-2.4.5-py3.4.egg\zc\buildout\buildout.py", line 770, in _develop
    zc.buildout.easy_install.develop(setup, dest)
  File "xxxx\.buildout\eggs\zc.buildout-2.4.5-py3.4.egg\zc\buildout\easy_install.py", line 985, in develop
    _detect_distutils_scripts(tmp3)
  File "xxxx\.buildout\eggs\zc.buildout-2.4.5-py3.4.egg\zc\buildout\easy_install.py", line 911, in _detect_distutils_scripts
    dev_script_content = fp.read()
  File "D:\bin\Python34\Lib\encodings\cp1251.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x98 in position 5011: character maps to <undefined>```

Version 2.4.4 doesn't have this issue.

**Comment by leorochael** [comment]
Is your package trying to distribute a binary executable file ("backend.exe") by (ab)using the [scripts](https://docs.python.org/2/distutils/setupscript.html#installing-scripts) key of `distutils`?

Notice from the linked documentation that only python files should be distributed this way, not arbitrary binaries. This is because the first line of the script is adjusted to point to the correct interpreter, which can't happen on a binary executable.

**Comment by reinout** [comment]
See https://github.com/buildout/buildout/blob/master/CHANGES.rst#245-2015-10-14

2.4.5 contains a fix that distutils-style scripts are now also generated for the package you're currently developing (`develop = .`).

And yes, those files are read because the first line needs to be changed.

**Comment by anton-tagunov** [comment]
@leorochael No, I don't use script section. backend.exe is result of entry_point. I have prepared minimal example to reproduce the problem: https://bitbucket.org/atagunov/buildout_exe_issue

**Comment by agroszer** [comment]
That breaks now winbot too...

http://winbot.zope.org/builders/zc_buildout_dev%20py_330_win64%20master/builds/1010/steps/dev.py/logs/stdio

Too bad that the output is eaten by buildbot

**Comment by reinout** [comment]
I think I understand the problem! The cause is #273, which I recently added.

When installing development eggs, it looks in the temp dir for distutils-style scripts that might need to be installed. It reads every file, looking for a certain piece of text. The `open()` command apparently chokes on binary files.

Binary files: ah, on windows a `.exe` is generated, too!

I suspect the issue is fixed if I add a simple `if filename.endswith('.exe'): continue` in there...

@anton-tagunov, can you check if that would work? The change is here: https://github.com/buildout/buildout/commit/27d24b0ec7360c9da0251cf92b567594a786bae0

Easiest for you is to quickly change it by hand in `d:\temp\buildout_exe_issue\eggs\zc.buildout-2.4.6-py3.4.egg\zc\buildout\easy_install.py`.

**Comment by reinout** [comment]
@agroszer is it easy for you to feed #280 to your buildbot now?

**Comment by agroszer** [comment]
@reinout YAY that fixed it

**Comment by anton-tagunov** [comment]
@reinout  I have applied your patch and it work as expected.
But I think your solution fixes not all problems.
For example, generated console script can include non-ascii symbols. In this case we will get the same exception. 

For example I have added initialization parameters:
```
[buildout]
develop = backend

parts = backend

[backend]
recipe = zc.recipe.egg:scripts
eggs =
  backend
initialization =
    print("русский текст☀☁☂")
```

The same exception is raised in this case (with your patch)

**Comment by mgedmin** [comment]
I'm now curious as to what would happen if I had a Python package using an old-style distutils script that had some non-ASCII characters in comments (say, ©), and the user used a different locale (or had the locale misconfigured causing a fallback to US-ASCII).

I'm not curious enough to actually test that case and file bugs.

**Comment by reinout** [comment]
@anton-tagunov is it the very same exception on the very same line? Such an `initialization =` line should have an effect on *writing* to a file. And we only write to files we know we must exist.

So... the file-reading code I fixed in #280 should have no effect on writing initialization code.

If writing initialization code goes wrong, I suspect the original distutils script file is already non-ascii or non-utf8 and there's a custom locale in place or so.

**Comment by reinout** [comment]
Ah, it got auto-closed because I merged the #280 pull request :-)

I just released 2.4.7, @anton-tagunov, can you check if that one works for you?

**Comment by anton-tagunov** [comment]
Sorry for confuse the exception is another. In this case the exception is raised during reading buildout.cfg.

Anyway I think that it is possible that file can be decoded incorrectly and we will get the exception.

I have tested 2.4.7 and it works. Thank you!
